### PR TITLE
[C&S]Adding LAI errors to om

### LINKED ISSUE #2071: [Crop & Soil] Log errors to OM instead of raising `ValueError`s in `LeafAreaIndex`
If values passed to a method are invalid in the `LeafAreaIndex` module, then they should be handled more elegantly than they are now.

## Status

If an invalid value is passed to `_determine_lai_shapes` in the `LeafAreaIndex` module, then a ValueError is thrown.

## Intention

Have the `LeafAreaIndex` module do a better job of handling invalid values, either by logging them to the OM before crashing or trying to find a fallback value, or both.

## Request

Before raising ValueErrors in `_determine_lai_shapes`, log the invalid values to the OM. Also, talk to SMEs to see if there are any values that could be used if an invalid value is found. If so, change the code to log an error then fallback on these values instead of crashing.

### PR DESCRIPTION
**Pull request by matthew7838**
Adds errors to OM when invalid values are used in determine_lai_shapes. 
### Context

Issue(s) closed by this pull request: closes #2071

### What

- Adds error to OM before throwing error and ending the simulation.
- Update unit tests.

### Why

For users to have more information when encountering error in the function `determine_lai_shapes`.

### Test plan

Run with updated uni tests.

**Note**
Since the values vary a lot from different types of crops to various stages of the crop, it is best not to continue the simulation with any preassumed values.

### CODE REVIEW

File: RUFAS/routines/field/crop/leaf_area_index.py [hunk 1]
```diff
@@ -229,15 +230,46 @@ def _determine_lai_shapes(
...
+            om.add_error(
+                "Invalid second heat fraction",
+                f"Second heat fraction should be greater than 0, got: {second_heat_fraction}.",
+                info_map,
+            )
             raise ValueError("second_heat_fraction must be greater than 0")
         if not 0 < first_leaf_fraction < 1:
+            om.add_error(
+                "Invalid first leaf fraction",
+                f"First leaf fraction must not be greater than 0 or less than 1, got: {first_leaf_fraction}.",
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/leaf_area_index.py]:
```suggestion
                f"First leaf fraction must be in the range (0, 1), got: {first_leaf_fraction}.",
```

File: RUFAS/routines/field/crop/leaf_area_index.py [hunk 2]
```diff
@@ -229,15 +230,46 @@ def _determine_lai_shapes(
...
+                info_map,
+            )
             raise ValueError("first_leaf_fraction must not be greater than 0 or less than 1")
         if not 0 < second_leaf_fraction < 1:
+            om.add_error(
+                "Invalid second leaf fraction",
+                f"Second_leaf_fraction must not be greater than 0 or less than 1, got: {second_leaf_fraction}.",
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/leaf_area_index.py]:
```suggestion
                f"Second_leaf_fraction must be in the range (0, 1), got: {second_leaf_fraction}.",
```

File: RUFAS/routines/field/crop/leaf_area_index.py [hunk 3]
```diff
@@ -229,15 +230,46 @@ def _determine_lai_shapes(
...
+                info_map,
+            )
             raise ValueError("second_leaf_fraction must not be greater than 0 or less than 1")
         if first_heat_fraction == second_heat_fraction:
+            om.add_error(
+                "Invalid first and second heat fraction combination.",
+                f"First_heat_fraction cannot be exactly equal to second_heat_fractions,"
+                f" got: first heat fraction: {first_heat_fraction} and second heat fraction: {second_heat_fraction}.",
```
Comment (Isgotkowitz) [inline RUFAS/routines/field/crop/leaf_area_index.py]:
```suggestion
                f"First_heat_fraction cannot be exactly equal to second_heat_fractions,"
                f" got first heat fraction equal to second heat fraction: {second_heat_fraction}.",
```

**Review by Isgotkowitz — APPROVED** [review]
Left some suggestions on the error messages, but otherwise looks good!

File: RUFAS/routines/field/crop/leaf_area_index.py [hunk 4]
```diff
@@ -229,15 +230,46 @@ def _determine_lai_shapes(
...
+            om.add_error(
+                "Invalid second heat fraction",
+                f"Second heat fraction should be greater than 0, got: {second_heat_fraction}.",
+                info_map,
+            )
             raise ValueError("second_heat_fraction must be greater than 0")
         if not 0 < first_leaf_fraction < 1:
+            om.add_error(
+                "Invalid first leaf fraction",
+                f"First leaf fraction must be in the range (0, 1), got: {first_leaf_fraction}.",
```
Comment (ew3361zh) [inline RUFAS/routines/field/crop/leaf_area_index.py]:
```suggestion
                f"'first_leaf_fraction' must be between 0 and 1 (exclusive), got: {first_leaf_fraction}.",
```

File: RUFAS/routines/field/crop/leaf_area_index.py [hunk 5]
```diff
@@ -229,15 +230,46 @@ def _determine_lai_shapes(
...
+                info_map,
+            )
             raise ValueError("first_leaf_fraction must not be greater than 0 or less than 1")
         if not 0 < second_leaf_fraction < 1:
+            om.add_error(
+                "Invalid second leaf fraction",
+                f"Second_leaf_fraction must be in the range (0, 1), got: {second_leaf_fraction}.",
```
Comment (ew3361zh) [inline RUFAS/routines/field/crop/leaf_area_index.py]: Same comment as for first leaf fraction. Should be explicit that it's not inclusive of the endpoints.

**Review by ew3361zh — APPROVED** [review]
A few suggestions to be more precise in what the error message is reporting but overall looks great!

**Pull request merged**